### DISCUSSION TITLE
GSoC 2024 Project 12: Prototype for JAX/Flax Models Support (or Keras 3, Mindspore)

### DISCUSSION THREAD

**Discussion by rkazants** [openvinotoolkit/openvino#23531] [Google Summer of Code]
Dear participants in GSoC'24,

Pay attention that [project idea 12](https://github.com/openvinotoolkit/openvino/wiki/Google-Summer-Of-Code) was extended with a choice to implement **Prototype for support of Flax/JAX models by OpenVINO**.

Feel free to ask your questions under this thread.

Best regards,
Roman

**Comment by rkazants** [comment]
@MonalSD, @awayzjj, @chux0519, @LucaTamSapienza, @AsakusaRinne, fyi

**Comment by SanftMonster** [comment]
Hi Roman,

I'm very interested in exploring this extended choice because I'm familiar with Tensorflow and XLA (As you can see, I'm one of the maintainers of [Tensorflow.NET](https://github.com/SciSharp/TensorFlow.NET)). Could you please talk about the rationale behind this idea more deeply? In the idea description, it's mentioned that "Jax/Flax provides much faster training than PyTorch/Tensorflow". However, I find this claim not entirely convincing because it's actually benefited from XLA compiler, which is unfair to be compared with native PyTorch code. (Please correct me if I'm mistaken here) 😊 I hope to be able to maintain this support for Jax/Flax in the long term if my prototype is accepted in the future, even after GSoC ends in September. So I would be more than happy if supporting Jax/Flax proves to be valuable for users and openvino team.

would you mind if I send you an email to ask for some suggestions directly? The proposal submission deadline is approaching, and I believe I can complete the draft this week. :) Additionally, should I aim to complete the work within 175 hours? I'm not sure if more time is necessary to ensure a thorough completion.

Thank you for your time and support.

Best regards,
Rinne

**Reply by rkazants** [reply]
Hi @AsakusaRinne,

The reason of having native support of JAX models by OpenVINO is that not all JAX models can be exported to SavedModel format as it is mentioned in release notes of JAX. I think it is the same story like PyTorch to ONNX format.
Regarding performance I saw several resources that mentioned this gap from PyTorch, here is one of them: https://discuss.pytorch.org/t/struggling-to-get-pytorch-fast-enough-to-use-in-public-competition/186015. They compare with `torch.compile()` scenario.

I think it will be honest if you send me official proposal with your suggestions for GSoC. Please ask other questions under this thread. You are warmly welcome to participate. And of course, stay with us to contribute after GSoC'24 finished.

Best regards,
Roman

**Reply by chux0519** [reply]
Hi Roman, Hi Rinne

Yesterday, I tried out Flax/JAX and looked into related materials. I'm quite interested in this area of implementation. However, my personal abilities are limited, so I might not apply for this idea. Nevertheless, I'm more than willing to participate in it, whether it's during the GSoC period or not.

I explored JAX2TF. Currently, the implementation involves compiling into a StableHLO, which can then be directly invoked in TensorFlow using relevant low-level APIs. Additionally, the serialized JAX model is essentially a TensorFlow op wrapped in an `XlaCallModule`, allowing it to be serialized into a SavedModel.

So, it seems that currently OpenVINO takes JAX2TF to SavedModel, then uses TensorFlow format before converting it into an OpenVINO model.

Additionally, I noticed that when using `flax.training.checkpoints.save_checkpoint_multiprocess`, the locally saved model is essentially a `pytree`. My understanding is that this `pytree` is more like a state, storing only data and not code. Therefore, after restoring the checkpoint, it's often necessary to instantiate relevant Flax classes (such as `nn.Module`) in order to perform inference. Directly parsing this pytree may not provide access to the ops required for the underlying computation graph.

My current feeling is that we might need to do some preprocessing on the Python side. For example, loading the model into JAX, then generating HLO. Finally, we would parse it at the HLO level and generate the OpenVINO model. I'm not sure if this approach is correct. Additionally, assuming it is correct and can be implemented. It's worth noting that this capability is not exclusive to JAX models. TensorFlow or PyTorch also possess this capability (although perhaps not fully supported). They can compile models into StableHLO, so theoretically, the same method could be used for conversion.

I am still very unfamiliar with the field of machine learning (compilers, models), so please do not hesitate to correct me if I am wrong about anything.

**Reply by SanftMonster** [reply]
Hi Roman @rkazants 

Thank you a lot for your response! I intend to send you a proposal in this weekend. Before I do, could you please confirm whether the project is supposed to be completed within 175 hours? In my opinion, additional time might be required to ensure comprehensive completion. However, I will adhere to the 175-hour timeline in my proposal if that is the final decision. :)

Additionally, would it be acceptable for me to submit proposals for two projects concurrently? Initially, my interest was in project 11, and I've finished the official proposal for it. However, this project with the prospect of integrating Jax/Flax support really attracts me  so much. Although I am more drawn to this project, I am inclined to continue to submit my proposal for project 11, given the substantial preparation I've invested. Thank you for your consideration!

Sincerely, 
Rinne

**Reply by rkazants** [reply]
Hi @chux0519,

You have the right exploration that we currently support JAX models by OV using JAX2TF. However, this tool has limitations and this is not direct and native support of JAX models. This approach requires intermediate model representation and serialization on a disk.

We need a sort of support for conversion of JAX/Flax object from memory, when you pass it into `convert_model`. Try to find other approaches.

Best regards,
Roman

**Reply by SanftMonster** [reply]
@chux0519 Hi, Yongsheng, 

I'm glad to see you here. I think you're right about that it's possible to get HLO IR from Jax/PyTorch/Tensorflow. Please refer to [pytorch-xla](https://github.com/pytorch/xla), [tf2xla](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/compiler/tf2xla) and [open-xla](https://github.com/openxla/xla). Initially xla is a part of tensorflow and it was extracted from it exactly because developers didn't want XLA (or HLO IR) could be used only in tensorflow. However, in OpenVINO, I don't think we should do similar things, to get HLO IR from models of all frameworks. Otherwise we're just repeating the same thing as open-xla does.

For GSoC application, I view our relationship as not only competitive but also collaborative. Though I'll submit a proposal for it, I'd like to share what I know and find during this preparation. I welcome any opportunity for us to discuss and exchange ideas freely. :)

Best regards, 
Rinne

**Reply by chux0519** [reply]
Understood. Actually, I'm more familiar with Stable Diffusion, and I've already applied for Proposal 4. However, the work discussed in this thread seems more interesting and challenging. I'll continue to keep an eye on it, and I believe that even after successfully implementing the proof of concept (POC) this summer, there will still be plenty of things to do. Additionally, Rinne, as far as I know, individuals are allowed to submit two proposals simultaneously (assuming you have the time). I look forward to exchanging ideas and learning together with everyone.

**Reply by SanftMonster** [reply]
@chux0519 Thank you for your reminder. I have previously reviewed the rules of GSoC, and my inquiry about acceptance is mainly to ask if mentors mind it, because only one proposal per student could be accepted finally. :)

I'm more than excited if having opportunity to work along with you on Jax/Flax support in the future! I will actively share my works with you if my proposal for this project passes.

**Comment by MonalSD** [comment]
Hey @rkazants , @AsakusaRinne and @chux0519 ,

It's truly invigorating to witness such a wealth of innovative ideas being brought to the table for this project. As a newcomer to the world of open source, I find this level of enthusiasm incredibly motivating. I agree with Yongsheng's point regarding the necessity of preprocessing on the Python side before parsing and generating the corresponding IR. Developing an adapter for JAX-based models to standardize them into an abstract format seems like a good approach to start with. This would facilitate seamless processing and compatibility moving forward.

Moreover, considering that locally saved models essentially adhere to a pytree structure, devising our reconstruction of this pytree during IR conversion would enable us to mirror the practices embraced by the wider community, thereby empowering new contributors with a familiar framework to engage with.

Additionally, I propose that benchmarking our converted OpenVINO models against the original JAX models could serve as a valuable tool. Not only would it provide us with insights into the potential impact we can achieve, but it would also bolster the credibility of our proof of concept.

Looking forward to delving deeper into these concepts with all of you.

**Comment by rkazants** [comment]
Dear contributors (@MonalSD, @awayzjj, @chux0519, @LucaTamSapienza, @AsakusaRinne),

If you want to apply for `Prototype for JAX/Flax Models Support (or Keras 3, Mindspore)`, please submit your final proposal through GSoC webapp (necessarily). The deadline for the final submission is 2nd April 6pm UTC.

Best regards,
Roman

**Reply by rkazants** [reply]
@Pranshu-S, please pay attention as well

**Reply by SanftMonster** [reply]
@rkazants I have sent my application to your email yesterday and I will appreciate it if you would like to give me some suggestions before I submit it to the GSoC portal. :)

**Reply by rkazants** [reply]
Hi @AsakusaRinne, I will take a look today later. Thanks!

**Reply by MonalSD** [reply]
Hi @rkazants I have mailed my application. It will be great to have your inputs.

**Reply by awayzjj** [reply]
@rkazants  Hi, I have sent the application to your email. I would greatly appreciate any feedback .
